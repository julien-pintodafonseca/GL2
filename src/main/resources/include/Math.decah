// Grenoble INP - Ensimag
// Projet GÃ©nie Logiciel 2020, Equipe GL2
// Library for class Math of Deca, coded in Deca

class Math {
    float MIN_VALUE = 1.4 * pow(10,-45);
    int MAX_EXPONENT = 127;
    int MIN_EXPONENT = -126;

    float ulp(float f) {
        int exp = getExponent(f);

        if (exp == MAX_EXPONENT+1) {
            return abs(f);
        } else if (exp == MIN_EXPONENT+1) {
            return MIN_VALUE;
        } else {
            println("ulp(f) not yet implemented");
        }
    }

    float sin(float f) {
        println("sin(f) not yet implemented");
    }

    float cos(float f) {
        println("cos(f) not yet implemented");
    }

    float asin(float f) {
        println("asin(f) not yet implemented");
    }

    float atan(float f) {
        println("atan(f) not yet implemented");
    }

    int getExponent(float x) {
        int e = 0;
        if (x == 0) {
            return 0;
        } else {
            while (pow(2,e) <= x) {
                e = e+1;
            }
            return e-1;
        }
    }

    float abs(float x) {
        if (x < 0) {
            return -x;
        } else {
            return x;
        }
    }

    float pow(float x, int e) {
        float p;
        if (e == 0) {
            return 1;
        } else {
            p = pow(x, e/2);

            if (e%2 == 0) {
                return p*p;
            } else if (e > 0) {
                return x*(p*p);
            } else {
                return (p*p)/x;
            }
        }
    }
}

// End of Deca Math library
