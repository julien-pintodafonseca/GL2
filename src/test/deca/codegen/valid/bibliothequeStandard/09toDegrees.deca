#include "Math.decah"

{
    Math m = new Math();

    println(m._toDegrees(0));                           // 0.00000e+00;
    println(m._toDegrees(0.0175));                      // 1.00000e+00;
    println(m._toDegrees(0.0873));                      // 5.00000e+00;
    println(m._toDegrees(0.1745));                      // 1.00000e+01;
    println(m._toDegrees(0.2618));                      // 1.50000e+01;
    println(m._toDegrees(0.3491));                      // 2.00000e+01;
    println(m._toDegrees(0.4363));                      // 2.50000e+01;
    println(m._toDegrees(0.5236));                      // 3.00000e+01;
    println(m._toDegrees(0.6109));                      // 3.50000e+01;
    println(m._toDegrees(0.6981));                      // 4.00000e+01;
    println(m._toDegrees(m._getPI()/4));                // 4.50000e+01;
    println(m._toDegrees(m._getPI()/3));                // 6.00000e+01;
    println(m._toDegrees(m._getPI()/2));                // 9.00000e+01;
    println(m._toDegrees(m._getPI()/1.5));              // 1.20000e+02;
    println(m._toDegrees(m._getPI()));                  // 1.80000e+02;
    println(m._toDegrees(m._getPI()*1.25));             // 2.25000e+02;
    println(m._toDegrees(m._getPI()*1.5));              // 2.70000e+02;
    println(m._toDegrees(m._getPI()*1.5+0.5236*2));     // 3.30000e+02;
    println(m._toDegrees(m._getPI()*2));                // 3.60000e+02;
    println(m._toDegrees(-0.2618));                     // -1.50000e+01;
    println(m._toDegrees(-(m._getPI()/2)));             // -9.00000e+01;
    println(m._toDegrees(-(m._getPI()/1.5)));           // -1.20000e+02;
    println(m._toDegrees(-(m._getPI())));               // -1.80000e+02;
    println(m._toDegrees(-(m._getPI()*1.25)));          // -2.25000e+02;
    println(m._toDegrees(-(m._getPI()*1.5)));           // -2.70000e+02;
    println(m._toDegrees(-(m._getPI()*1.5+0.5236*2)));  // -3.30000e+02;
}
