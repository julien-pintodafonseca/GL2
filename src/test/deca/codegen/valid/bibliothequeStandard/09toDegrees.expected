0.00000e+00
1.00268e+00
5.00192e+00
9.99811e+00
1.50000e+01
2.00020e+01
2.49981e+01
3.00001e+01
3.50020e+01
3.99982e+01
4.50000e+01
6.00000e+01
9.00000e+01
1.20000e+02
1.80000e+02
2.25000e+02
2.70000e+02
3.30000e+02
3.60000e+02
-1.50000e+01
-9.00000e+01
-1.20000e+02
-1.80000e+02
-2.25000e+02
-2.70000e+02
-3.30000e+02
